### PR TITLE
DSDEEPB-1310 "Big hammer" workspace locking

### PR DESCRIPTION
**Pull request by helgridly**
- [x] **Submitter**: Rebase to master. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] **Submitter**: **Assign PR to** 
- [x] : does a once-over of the PR
- [x] : Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  * Rest of team may comments on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter updates documentation as needed
  * Submitter rebases to master again if necessary
  * Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x] **LR**: sign off, **assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to master
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: Verify swagger UI on dev server still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala [hunk 1]
```diff
@@ -99,6 +100,79 @@ class DataSource(graphFactory: OrientGraphFactory) extends LazyLogging {
...
+  private def releaseWriteLock(elem: Element) = {
+    assert(writeLocks(elem) > 0, s"Attempting to release nonexistent write lock on $elem")
+    writeLocks(elem) -= 1
+    if( writeLocks(elem) == 0 ) {
+      elem.asInstanceOf[OrientElement].unlock()
+    }
+  }
+
+  def withWriteLock[T](elem: Element) (op: => T): T = {
+    acquireWriteLock(elem)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala]: try/finally?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala [hunk 2]
```diff
@@ -99,6 +100,79 @@ class DataSource(graphFactory: OrientGraphFactory) extends LazyLogging {
 class RawlsTransaction(val graph: OrientGraph, dataSource: DataSource) {
   private val rollbackOnly = new AtomicBoolean(false)
 
+  var readLocks:mutable.Map[Element, Int] = mutable.Map[Element, Int]().withDefaultValue(0)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala]: is mutable.Map thread-safe?

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala]: I meant to switch this over to `scala.collection.mutable.SynchronizedMap` after the PR comments the last time around. Thanks for the catch.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -342,8 +342,8 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, containerDAO:
 
   def copyEntities(entityCopyDef: EntityCopyDefinition, uri: Uri): Future[PerRequestMessage] =
     dataSource inFutureTransaction { txn =>
-      withWorkspaceContextAndPermissions(entityCopyDef.sourceWorkspace, WorkspaceAccessLevel.Read, txn) { sourceWorkspaceContext =>
-        withWorkspaceContextAndPermissions(entityCopyDef.destinationWorkspace, WorkspaceAccessLevel.Write, txn) { destWorkspaceContext =>
+      withWorkspaceContextAndPermissions(entityCopyDef.destinationWorkspace, WorkspaceAccessLevel.Write, txn) { destWorkspaceContext =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: add comment that order of withWorkspaceContextAndPermissions is important here and why

**Comment by dvoet** [comment]
add comments where lock ordering is important then :+1:

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphEntityDAOSpec.scala [hunk 4]
```diff
@@ -240,7 +240,7 @@ class GraphEntityDAOSpec extends FlatSpec with Matchers with OrientDbTestFixture
 
   it should "list entities" in withDefaultTestDatabase { dataSource =>
     dataSource.inTransaction { txn =>
-      withWorkspaceContext(testData.workspace, txn) { context =>
+      withWorkspaceContext(testData.workspace, writeLock = true, txn) { context =>
```
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphEntityDAOSpec.scala]: why does listing entities require a write lock?

**Pull request merged**